content/5_Plat/PIE.mdx

```diff
@@ -0,0 +1,284 @@
+	if (mobius[i]) {
+		mobius[i] = -mobius[i];
+		for (int j = 2 * i; j < VALMAX; j += i) { mobius[j] += mobius[i]; }
+	}
+}
+```
+
+</CPPSection>
+</LanguageSection>
+
+## Applications
+
+### SQFREE
+
+<FocusProblem problem="SQFREE" />
+
+A perfect application for inclusion-exclusion principle and mobius function. In this particular case the set $A_i$ - previously mentioned in the tutorial section - denotes how many numbers are numbers are divisible with $i^2$ and we're asked to find out $\bigg| \bigcup_{i=1}^{\sqrt{n}} A_i \bigg|$. the precomputed mobius array tells whether to add or substract $A_i$.
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <iostream>
+#include <vector>
+
+using namespace std;
+
+const int VALMAX = 2e7;
+
+int mobius[VALMAX];
+
+int main() {
+	int test_num;
+	cin >> test_num;
+
+	mobius[1] = -1;
+	for (int i = 1; i < VALMAX; i++) {
+		if (mobius[i]) {
+			mobius[i] = -mobius[i];
+			for (int j = 2 * i; j < VALMAX; j += i) { mobius[j] += mobius[i]; }
+		}
+	}
+
+	for (int t = 0; t < test_num; t++) {
+		long long n;
+		long long ans = 0;
+		cin >> n;
+		for (int i = 1; 1LL * i * i <= n; i++) {
+			ans += mobius[i] * n / ((long long)i * i);
+		}
+		cout << ans << '\n';
+	}
+}
+```
+
+</CPPSection>
+</LanguageSection>
+
+### Cowpatibility
+
+<FocusProblem problem="cow" />
+
+In this particular case the set $A_i$ - previously mentioned in the tutorial section - denotes how many pairs of cows have at least $i$ ice cream flavors in common. From the total number of pairs substract the union of $A_i$. The global answer is:
+$$
+\frac{n \cdot(n-1)}{2}- \bigg| \bigcup_{i=1}^{5} A_i \bigg|
+$$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+
+using namespace std;
+
+int main() {
+	ifstream in("cowpatibility.in");
+	int n;
+	in >> n;
+
+	map<vector<int>, int> subsets;
+	for (int i = 1; i <= n; i++) {
+		vector<int> v(5);
+		for (int &num : v) { in >> num; }
+		sort(v.begin(), v.end());
+		for (int mask = 1; mask < (1 << 5); mask++) {
+			vector<int> subset;
+			for (int i = 0; i < 5; i++) {
+				if ((1 << i) & mask) { subset.push_back(v[i]); }
+			}
+			subsets[subset]++;
+		}
+	}
+
+	long long ans = (long long)n * (n - 1) / 2;
+	for (const auto &[subset, freq] : subsets) {
+		ans -= ((int)subset.size() % 2 == 1 ? 1LL : -1LL) * freq * (freq - 1) / 2;
+	}
+
+	ofstream("cowpatibility.out") << cnt << '\n';
+}
+```
+
+</CPPSection>
+</LanguageSection>
+
+### The number of strings that match a certain pattern
+
+<FocusProblem problem="patterns" />
+
+A dynamic programming approach with bitmasking would look like this: $dp[i][mask] = \text{the number of strings of length i that match all the patterns in set, but none other patterns. } $ The recurrence is:
+$$
+dp[i][mask \& j]=dp[i-1][j]\text{ where j is a set of patterns that match charachter c at position i}
+$$
+The following code illustrates this:
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+int howMany(vector<string> patterns, int k) {
```

Comment:
is the camelcase name necessary idk lol

Reply:
That the name of the function on topcoder, otherwise get AC.

https://community.topcoder.com/stat?c=problem_statement&pm=8307

Reply:
yeah that's what i thought